server/src/main/java/org/elasticsearch/repositories/blobstore/ChunkedBlobOutputStream.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+package org.elasticsearch.repositories.blobstore;
+
+import org.elasticsearch.common.io.stream.ReleasableBytesStreamOutput;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.core.Releasables;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Base class for doing chunked writes to a blob store. Some blob stores require either up-front knowledge of the size of the blob that
+ * will be written or writing it in chunks that are then joined into the final blob at the end of the write. This class provides a basis
+ * on which to implement an output stream that encapsulates such a chunked write.
+ *
+ * @param <T> type of chunk identifier
+ */
+public abstract class ChunkedBlobOutputStream<T> extends OutputStream {
```

Comment:
I think we could simplify this class a bit and encapsulate implementation details

i.e: 
```java
protected abstract T writePart(ReleasableBytesStreamOutput buffer) throws IOException;
protected abstract void writeEntireBuffer(ReleasableBytesStreamOutput buffer) throws IOException;

public void flushBuffer() {
    try {
        T partId = writePart(buffer);
        finishPart(partId);
    } catch (Exception e) {
        failed = true;
        handleFailure();
        throw e;
    }
}

public void close() {
    if (failed) {
        // cleanup
    } else {
       // handle last part commit  || send entire buffer
    }
}
```

This way, implementors only need to take care of implementing `writePart` and `writeEntireBuffer`. But it's just a suggestion 👍

Reply:
I thought about this a little now and I agree this looks nicer. The problem with it though is that it does not account for exceptions other than those from writing to the blob store (e.g. serialization exceptions). So e.g. if you write out `RepositoryData` and `toXContent` throws some NPE or so, you'd write a broken blob without additional precautions => I think we need to expose part of the failure handling to the caller because this must also handle caller exceptions right?

Reply:
You're right, I think we should keep the `markSucess()` method to signal a successful execution